mlflow/tracking/context.py

```diff
@@ -0,0 +1,187 @@
+    return "<console>"
+
+
+def _get_git_commit(path):
+    try:
+        import git
+    except ImportError as e:
+        _logger.warning(
+            "Failed to import Git (the Git executable is probably not on your PATH),"
+            " so Git SHA is not available. Error: %s", e)
+        return None
+    try:
+        if os.path.isfile(path):
+            path = os.path.dirname(path)
+        repo = git.Repo(path, search_parent_directories=True)
+        commit = repo.head.commit.hexsha
+        return commit
+    except (git.InvalidGitRepositoryError, git.GitCommandNotFound, ValueError, git.NoSuchPathError):
+        return None
+
+
+def _get_source_version():
+    main_file = _get_main_file()
+    if main_file is not None:
+        return _get_git_commit(main_file)
+    return None
+
+
+def _get_source_type():
+    return SourceType.LOCAL
+
+
+class ContextProvider(object):
+
+    __metaclass__ = ABCMeta
+
+    @abstractmethod
+    def in_context(self):
+        """
+        Determine if MLflow is running in this context.
+
+        :return: bool indicating if in this context
+        """
+        pass
+
+    @abstractmethod
+    def tags(self):
+        """
+        Generate context-specific tags.
+
+        :return: dict of tags
+        """
+        pass
+
+
+class DefaultContext(ContextProvider):
+    def in_context(self):
+        return True
+
+    def tags(self):
+        return {
+            MLFLOW_SOURCE_NAME: _get_source_name(),
+            MLFLOW_SOURCE_TYPE: _get_source_type()
+        }
+
+
+class GitContext(ContextProvider):
+    def in_context(self):
+        return _get_source_version() is not None
+
+    def tags(self):
+        return {
+            MLFLOW_GIT_COMMIT: _get_source_version()
+        }
+
+
+class DatabricksNotebookContext(ContextProvider):
+    def in_context(self):
+        return databricks_utils.is_in_databricks_notebook()
+
+    def tags(self):
+        notebook_id = databricks_utils.get_notebook_id()
+        notebook_path = databricks_utils.get_notebook_path()
+        webapp_url = databricks_utils.get_webapp_url()
+        tags = {
+            MLFLOW_SOURCE_NAME: notebook_path,
+            MLFLOW_SOURCE_TYPE: SourceType.NOTEBOOK
+        }
+        if notebook_id is not None:
+            tags[MLFLOW_DATABRICKS_NOTEBOOK_ID] = notebook_id
+        if notebook_path is not None:
+            tags[MLFLOW_DATABRICKS_NOTEBOOK_PATH] = notebook_path
+        if webapp_url is not None:
+            tags[MLFLOW_DATABRICKS_WEBAPP_URL] = webapp_url
+        return tags
+
+
+class ContextProviderRegistry(object):
+    """Registry for context provider implementations
+
+    This class allows the registration of a context provider which can be used to infer meta
+    information about the context of an MLflow experiment run. Implementations declared though the
+    entrypoints `mlflow.context_provider` group can be automatically registered through the
+    `register_entrypoints` method.
```

Comment:
Ordering here matters, so we should probably document it. I suppose it would be sufficient to say that registered plugins are allowed to override tags in the base context provider (but that the order between registered plugins is non-deterministic).